File: engine/internal/game/behaviour/movement.go
package behaviour

import (
	"github.com/goofr-group/game-engine/pkg/action"
	"github.com/goofr-group/game-engine/pkg/engine"
	"github.com/goofr-group/go-math/mathf"
	"github.com/goofr-group/go-math/rotation/matrix"
	"github.com/goofr-group/physics-engine/pkg/game"

	"github.com/goofr-group/jump-master/engine/internal/config"
	input "github.com/goofr-group/jump-master/engine/internal/game/action"
	"github.com/goofr-group/jump-master/engine/internal/game/animation"
	"github.com/goofr-group/jump-master/engine/internal/game/property"
)

// Movement defines the structure of the movement behaviour.
type Movement struct {
	object        *game.Object
	actionManager *action.Manager
	config        config.Movement

	checkGround *CheckGround
	animator    *Animator

	leftAction  bool
	rightAction bool
	jumpAction  bool
}

// NewMovement returns a new movement behaviour with the given configuration.
func NewMovement(
	object *game.Object,
	actionManager *action.Manager,
	config config.Movement,
	checkGround *CheckGround,
	animator *Animator,
) Movement {
	return Movement{
		object:        object,
		actionManager: actionManager,
		config:        config,
		checkGround:   checkGround,
		animator:      animator,
	}
}

func (b Movement) Enabled() bool {
	return true
}

func (b *Movement) FixedUpdate(e *engine.Engine) error {
	time := e.Time()

	// Check if the rigid body is accessible.
	if b.object == nil {
		return nil
	}
	if b.object.RigidBody == nil {
		return nil
	}

	// Check if the object is in contact with the ground.
	if !b.checkGround.Grounded() || b.object.RigidBody.Velocity.Y > Epsilon {
		return nil
	}

	// Check if the jump action is being performed.
	if b.jumpAction {
		if b.leftAction {
			b.object.SetProperty(property.FlipHorizontally, true)
		}
		if b.rightAction {
			b.object.SetProperty(property.FlipHorizontally, false)
		}
		return nil
	}

	// Compute the direction to add to the object based on the left and right actions.
	var direction float64
	if b.leftAction {
		// Add the left direction.
		direction -= 1
		b.object.SetProperty(property.FlipHorizontally, true)
	}
	if b.rightAction {
		// Add the right direction.
		direction += 1
		b.object.SetProperty(property.FlipHorizontally, false)
	}

	// Reset the horizontal velocity of the object when no movement action is performed.
	if mathf.Approximately(direction, 0) {
		b.object.RigidBody.Velocity.X = 0
		b.animator.SetAnimation(animation.Idle)
		return nil
	}

	// Add the computed velocity when the movement actions are performed.
	b.object.RigidBody.Velocity.X = direction * b.config.Speed * time.FixedDeltaTime
	b.animator.SetAnimation(animation.Walk)

	return nil
}

func (b *Movement) Update(_ *engine.Engine) error {
	// Avoid the object from rotating.
	b.object.Transform.Rotation = matrix.Identity()

	// Update the actions.
	b.leftAction = b.actionManager.Action(input.Left)
	b.rightAction = b.actionManager.Action(input.Right)
	b.jumpAction = b.actionManager.Action(input.Jump)

	return nil
}

File: Assets/Scripts/Movement.cs
using UnityEngine;
using UnityEngine.InputSystem;

public class Movement : MonoBehaviour
{
    /// <summary>
    /// Defines the movement speed multiplier.
    /// </summary>
    [SerializeField] private float speed;

    // Action status.
    private bool leftActionStatus;
    private bool jumpActionStatus;
    private bool rightActionStatus;

    // Required components.
    private SpriteRenderer spriteRenderer;
    private new Rigidbody2D rigidbody2D;
    private CheckGround checkGround;

    // Input actions.
    private InputAction moveAction;
    private InputAction jumpAction;

    private void Awake()
    {
        spriteRenderer = GetComponent<SpriteRenderer>();
        rigidbody2D = GetComponent<Rigidbody2D>();
        checkGround = GetComponentInChildren<CheckGround>();

        moveAction = InputSystem.actions.FindActionMap(GameManager.INPUT_ACTION_MAP_PLAYER).FindAction(GameManager.INPUT_ACTION_PLAYER_MOVE);
        jumpAction = InputSystem.actions.FindActionMap(GameManager.INPUT_ACTION_MAP_PLAYER).FindAction(GameManager.INPUT_ACTION_PLAYER_JUMP);
    }

    private void Update()
    {
        // Update the actions.
        float moveValue = moveAction.ReadValue<float>();

        leftActionStatus = moveValue < 0;
        rightActionStatus = moveValue > 0;
        jumpActionStatus = jumpAction.IsPressed();
    }

    private void FixedUpdate()
    {
        // Check if the object is in contact with the ground.
        if (!checkGround.IsTouchingGround() || rigidbody2D.linearVelocityY > Mathf.Epsilon)
        {
            return;
        }

        // Compute the direction to add to the object based on the left and right actions.
        int direction = 0;
        if (leftActionStatus)
        {
            // Add the left direction.
            direction -= 1;
            spriteRenderer.flipX = true;
        }
        if (rightActionStatus)
        {
            // Add the right direction.
            direction += 1;
            spriteRenderer.flipX = false;
        }

        // Reset the horizontal velocity of the object 
        // when no movement action is performed or the jump action is being performed.
        if (direction == 0 || jumpActionStatus)
        {
            rigidbody2D.linearVelocityX = 0;
            return;
        }

        // Add the computed velocity when the movement actions are performed.
        rigidbody2D.linearVelocityX = direction * speed * Time.fixedDeltaTime;
    }
}